protocols/raft/src/main/java/io/atomix/protocols/raft/storage/RaftStorage.java

```diff
@@ -227,6 +229,37 @@ public StorageStatistics statistics() {
     return statistics;
   }
 
+  /**
+   * Attempts to acquire a lock on the storage directory.
+   *
+   * @param id the ID with which to lock the directory
+   * @return indicates whether the lock was successfully acquired
+   */
+  public boolean lock(String id) {
```

Comment:
what about something like this? https://stackoverflow.com/a/9020391/7534815

Reply:
Not necessary in this case. It's overkill. We're not trying to control concurrent access to a shared file. All we need is to ensure only a single file can operate on a log directory, and this accomplishes that. The `File.createNewFile` call atomically creates a file. If the file does not exist, the first node to create the file will effectively acquire a persistent, forever lock on the data directory. If it does exist, the node just must read the file to verify that it's the one that created it in the past. There's no need to use channel locks because this lock is never actually released for the purposes of allowing another node to acquire it.

Reply:
We could release it immediately, point was to address the race you mentioned.